spring-rabbit-test/src/test/java/org/springframework/amqp/rabbit/test/ExampleRabbitListenerTest.java

```diff
@@ -0,0 +1,158 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.amqp.rabbit.test;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertThat;
+
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import org.springframework.amqp.core.AnonymousQueue;
+import org.springframework.amqp.core.Queue;
+import org.springframework.amqp.rabbit.annotation.RabbitListener;
+import org.springframework.amqp.rabbit.config.SimpleRabbitListenerContainerFactory;
+import org.springframework.amqp.rabbit.connection.CachingConnectionFactory;
+import org.springframework.amqp.rabbit.connection.ConnectionFactory;
+import org.springframework.amqp.rabbit.core.RabbitAdmin;
+import org.springframework.amqp.rabbit.core.RabbitTemplate;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.messaging.handler.annotation.Header;
+import org.springframework.messaging.handler.annotation.Payload;
+import org.springframework.test.annotation.DirtiesContext;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+
+/**
+ * @author Gary Russell
+ * @since 1.6
+ *
+ */
+@ContextConfiguration
+@RunWith(SpringJUnit4ClassRunner.class)
+@DirtiesContext
+public class ExampleRabbitListenerTest {
+
+	@Rule
+	public BrokerRunning brokerRunning = BrokerRunning.isRunning();
+
+	@Autowired
+	private RabbitTemplate rabbitTemplate;
+
+	@Autowired
+	private Queue queue1;
+
+	@Autowired
+	private Queue queue2;
+
+	@Autowired
+	private RabbitListenerTestHarness harness;
+
+	@Test
+	public void testTwoWay() throws Exception {
+		assertEquals("FOO", this.rabbitTemplate.convertSendAndReceive(this.queue1.getName(), "foo"));
+
+		BlockingQueue<Object[]> argumentQueue = this.harness.getArgumentQueueFor("foo");
+		assertNotNull(argumentQueue);
+		Object[] args = argumentQueue.poll(10, TimeUnit.SECONDS);
+		assertThat((String) args[0], equalTo("foo"));
+		BlockingQueue<Object> resultQueue = this.harness.getResultQueueFor("foo");
+		assertNotNull(resultQueue);
+		Object result = resultQueue.poll(10, TimeUnit.SECONDS);
```

Comment:
Wouldn't it be more friendlier to have something like:
```
ArgumentsAndResult argumentsAndResult = this.harness.getArgumentsAndResultFor("foo");
ArgumentsAndResult argumentsAndResult = this.harness.getArgumentsAndResultFor("foo", 10, TimeUnit.SECONDS);
```
?
To avoid the `BlockingQueue` noise.